### PR TITLE
Home page issue #31

### PR DESCRIPTION
**Pull request by bxchang04**
Successful for the most part. However, there are several inconsistencies with the old and new CRWA website templates.

Please see comments for more details. This will most likely be my last pull request until I finish my Java training by mid-Nov. It's been great working with y'all...looking to be back soon.

### CODE REVIEW

**Comment by dwreeves** [comment]
Expanding the dicts to include multiple key-value stores within the reach keys is a great approach to the problem, in my opinion. I'd have done something very similar to what you're doing.

There are a few things we can do to optimize the code (so we don't run the model functions two times per reach) and a few things (we want to use lists instead of dicts for the reaches, since we aren't using the key for anything). I'll provide more notes momentarily after I run your code and look at it in more depth.

**Comment by dwreeves** [comment]
Just ran this. The table design is very good and that's exactly what I had in mind!

File: flagging_site/blueprints/flagging.py [hunk 1]
```diff
@@ -72,13 +72,57 @@ def index() -> str:
...
+        5: {
+            'flag': reach_5_model(df, rows=1)['safe'].iloc[0],
+            'time': reach_5_model(df, rows=25)['time'].iloc[0],
+            'names': {
+                0: 'Charles River Yacht Club',
+                1: 'Union Boat Club',
+                2: 'Community Boating',
+                3: 'CRCK at Kendall Square', # Found on old CRWA page only
+                4: 'Paddle Boston at Kendall Square' # Assuming this is reach 5
+            }
```
Comment (dwreeves) [inline flagging_site/blueprints/flagging.py:120]: Make all of these lists instead of dicts.

Also, instead of `names`, call it `boathouses`. "Name" is vague (name of what, exactly?), but "boathouses" is not.

For example, rows 114-120 would be updated to:

```python
            'boathouses': [
                'Charles River Yacht Club',
                'Union Boat Club',
                'Community Boating',
                'CRCK at Kendall Square', # Found on old CRWA page only
                'Paddle Boston at Kendall Square' # Assuming this is reach 5
            ]
```

So, why lists and not dicts? The reason why is because a dict is a key-value store; it maps a key to a specific value. However, here the keys `0`, `1`, etc. have no special significance in this case. Charles River Yacht Club, for example, has no special significance in its relation to `0` other than the arbitrary choice to put it first and to give it a 0. You'd rather simply use a list in this case.

The list will preserve the order when being passed into Jinja, so don't worry about order. In fact, list order is safer to preserve than dict order (for example, in version of Python before 3.5, dicts were totally randomly ordered, whereas lists were not).

BTW, good work adding in those comment! I can definitely ask CRWA about those ones you're not sure of.

File: flagging_site/templates/index.html [hunk 2]
```diff
@@ -1,30 +1,82 @@
...
+    </tr>
     {% endfor %}
-
+</table>
+{% for key, value in homepage.items() %}
+<p class="home-centered">
+    <!-- Currently uses time of last row for reach 5 table. Needs unit test to ensure this matches the times for reach 2, 3, and 4 -->
+    {% if loop.last %}
+    (Last updated: {{value.time}})
+    {% endif %}
```
Comment (dwreeves) [inline flagging_site/templates/index.html:79]: Wow! This is a very astute observation.

You are basically pointing to some of the limitations of the current code base, and you're doing a good job dealing with those limitations.

Your note about unit-testing this is totally correct and not something I had considered yet as part of unit-testing for the data. Good job! I'll make sure the unit-test writers for the data models include this as part of their tests.

For now, I know that the models all have the same final timestamp. It's something we need to test though, or guarantee in some other way.

Also for now, my suggestion is that we'll pass only _one_ time value in, since we're only using one anyway to begin with. Let's do this for now:

`(Last updated: {{model_last_updated_time}}`

File: flagging_site/blueprints/flagging.py [hunk 3]
```diff
@@ -72,13 +72,57 @@ def index() -> str:
...
+        }
     }
-    return render_template('index.html', flags=flags)
+
+    return render_template('index.html', homepage=homepage)
```
Comment (dwreeves) [inline flagging_site/blueprints/flagging.py:124]: Add `model_last_updated_time=model_last_updated_time`, in combination with the comment I make about the `time` keys.

File: flagging_site/blueprints/flagging.py [hunk 4]
```diff
@@ -72,13 +72,57 @@ def index() -> str:
...
+        4: {
+            'flag': reach_4_model(df, rows=1)['safe'].iloc[0],
+            'time': reach_4_model(df, rows=25)['time'].iloc[0],
+            'names': {
+                0: 'Riverside Boat Club'
+            }
+        },
+        5: {
+            'flag': reach_5_model(df, rows=1)['safe'].iloc[0],
+            'time': reach_5_model(df, rows=25)['time'].iloc[0],
```
Comment (dwreeves) [inline flagging_site/blueprints/flagging.py:113]: Remove the `time` key from each dict and somewhere in this function block, simply do this:

```python
model_last_updated_time = reach_5_model(df, rows=1)['time'].iloc[0]
```

As a note, The way you're getting the time currently doesn't work. The reason why is subtle, but it's a combination of how `rows=?` works, how the `reach_#_model` functions work, and how Pandas's `iloc` works.

So basically, you're defining `rows=25`. This means we're returning 25 rows of the model. The rows of the model are in ascending order by time. This means the _oldest_ time is at the top of the dataframe, and the _newest_ time is at the bottom of the dataframe.

`iloc[0]` returns the first row of the data. Do you see the problem yet? If you have 25 rows and the oldest data is at the top, you're returning the timestamp from 25 hours ago!

`rows=1` solves this problem because when `rows=1`, then the _only_ row is also the first row, and it also happens to be the _latest_ row too.

File: flagging_site/templates/index.html [hunk 5]
```diff
@@ -1,30 +1,82 @@
...
+    CRWA estimates E. coli bacteria levels using models that take into account a variety of environmental factors
+    including rainfall and river flow. CRWA collects water samples weekly to verify the model forecasts. Cyanobacteria
+    monitoring is conducted every 1-2 weeks.
+</p>
+<br />
+<!-- Is this needed? CRWA's website links to a 2019 report using a button 
+<p>
+    Subscribe to water quality alerts through the sign up form below.
+</p> -->
+<!-- optional 2019 report button here? -->
```
Comment (dwreeves) [inline flagging_site/templates/index.html:32]: I agree, this probably isn't needed.

As a note, the issue with HTML comments `<!-- comment here -->` is that they render in the HTML. But usually you don't want that. I would instead preference utilizing Jinja comment styles: `{# comment here #}` because these do not render in the HTML. If we were working in PHP instead of Jinja, the same general concept would be true: you'd preference `/* comment here */` inside PHP blocks over using the HTML comments.

Basically, it's preferable to not render comments in production. Except in circumstances like, for example in a lot of modern JS framework such as React and Angular, HTML comments are utilized as nodes within the document's object model and thus have a functional purpose inside of the rendered page. But that's totally different than a real, bona fide comment.
